frame/dapps-staking/src/lib.rs

```diff
@@ -97,6 +130,128 @@ impl Default for Version {
     }
 }
 
+#[derive(Encode, Decode, Clone, Copy, PartialEq, Eq, RuntimeDebug, TypeInfo)]
+pub struct EraStake<Balance: AtLeast32BitUnsigned + Copy> {
+    /// Staked amount in era
+    staked: Balance,
+    /// Staked era
+    era: EraIndex,
+}
+
+impl<Balance: AtLeast32BitUnsigned + Copy> EraStake<Balance> {
+    fn new(staked: Balance, era: EraIndex) -> Self {
+        Self { staked, era }
+    }
+}
+
+/// Contains information about how much was staked in each era
+#[derive(Encode, Decode, Clone, Default, PartialEq, Eq, RuntimeDebug, TypeInfo)]
+pub struct StakerInfo<Balance: AtLeast32BitUnsigned + Copy> {
+    // Size of this list would be limited by a configurable constant TODO
+    stakes: Vec<EraStake<Balance>>,
+}
+
+impl<Balance: AtLeast32BitUnsigned + Copy> StakerInfo<Balance> {
+    /// true if no stakes exist, false otherwise
+    fn is_empty(&self) -> bool {
+        self.stakes.is_empty()
+    }
+
+    /// number of `EraStake` chunks
+    fn len(&self) -> u32 {
+        self.stakes.len() as u32
+    }
+
+    /// Stakes some value in the specified era.
+    /// User should ensure that given era is either equal or greater than the
+    /// latest available era in the staking info.
+    fn stake(&mut self, current_era: EraIndex, value: Balance) -> Result<(), &str> {
+        if self.stakes.is_empty() {
+            self.stakes.push(EraStake::new(value, current_era))
+        } else {
+            let era_stake = self.stakes.last().unwrap(); // exists if vec not empty
+            if era_stake.era > current_era {
+                return Err("Unexpected era".into());
+            }
+
+            let new_stake_value = era_stake.staked.saturating_add(value);
+
+            if current_era == era_stake.era {
+                *self.stakes.last_mut().unwrap() = EraStake::new(new_stake_value, current_era)
+            } else {
+                self.stakes
+                    .push(EraStake::new(new_stake_value, current_era))
+            }
+        }
+        Ok(())
+    }
+
+    /// Unstakes some value in the specified era.
+    /// User should ensure that given era is either equal or greater than the
+    /// latest available era in the staking info.
+    fn unstake(&mut self, current_era: EraIndex, value: Balance) -> Result<(), &str> {
+        if self.stakes.is_empty() {
+            return Ok(());
+        }
+
+        let era_stake = self.stakes.last().unwrap(); // not empty so it exists
+        if era_stake.era > current_era {
+            return Err("Unexpected era".into());
+        }
+
+        let new_stake_value = era_stake.staked.saturating_sub(value);
+
+        if current_era == era_stake.era {
+            *self.stakes.last_mut().unwrap() = EraStake::new(new_stake_value, current_era)
+        } else {
+            self.stakes
+                .push(EraStake::new(new_stake_value, current_era))
+        }
```

Comment:
code duplication? same is in stake()

Reply:
Good catch!
I tried to avoid this by wrapping it up in a method but the problem is I cannot make it private.
And I don't want to expose this as part of the API.

E.g. it would be something like 'update_staked_value(era, value)`